### DISCUSSION TITLE
SAMD21 Xplained Pro micropython

### DISCUSSION THREAD

**Discussion by AlessandroCortese** [micropython/micropython#13790] [SAMD]
Hi, i would like to use SAMD21 Xplained Pro board with micropython. 
Do you know if i can access to all the UARTs port available in sad21 microcontroller?

**Comment by robert-hh** [comment]
You can access the UART ports of the board, as long as the pins are physically accessible. See the pinout documentation at https://docs.micropython.org/en/latest/samd/pinout.html#samd21-xplained-pro-pin-assignment-table
The `Serial` columns denote the UART (and I2C/SPI) pins in the notation port/pad. TX is at a pad 0 or 2, RX at any other pad. For pins 34/35 for example, the values are 5/0 for pin 34 and 5/1 for pin 35. telling that UART 5 TX is at pin 34, UART 5 RX is at Pin 35. Glancing through the table it looks as if all of UART 0 to 5 are available.

**Reply by AlessandroCortese** [reply]
Thanks for your reply Robert.
"TX is at a pad 0 or 2, RX at any other pad. " means that i can use UART5 tx both for pin 34 and for 32 (PB00) and UART rx both for pin 35 and for 33 (PB01)?

**Reply by robert-hh** [reply]
Yes, that's how I understand the data sheet. At a ItsyBitsyM0 for instance the default TX and RX are at pads 2 and 3.

**Reply by AlessandroCortese** [reply]
ok perfect. I have just bought samd21 xplained pro and followed instructions (https://micropython.org/download/SAMD21_XPLAINED_PRO/) but without success. I am connecting to Debug Usb port, i can see a virtual drive, i drag the file v1.22.2 (2024-02-22) .uf2 inside it but still remain the drive (i think it doesn't install). suggestions?
i have never faced problems with raspberry pico or esp32, very strange

**Reply by robert-hh** [reply]
Hard  to tell. I never had that board. Does the .uf2 file disappear from the board drive after a few seconds you have copied it?

**Reply by robert-hh** [reply]
Did you try the target USB interface? That's the one used by MicroPython.

**Reply by AlessandroCortese** [reply]
No for the Moment only debug port buy the strange things Is that the file does not disappear After copying it. Which interpreter do.you use for micropython and thonny as ide? Micropython generic?

**Reply by robert-hh** [reply]
According to the schematics, the debug port is not connected directly to the SAMD21. So you have to use the target port for communication with the SAMD. In combination with an Debugger like Segger Ozone you could use the debug port to upload an .elf file. You could try to upload an .hex file with the debug port. Some debug ports allow that. But both the .elf and .hex file are not provided by default.

For communication with the board I use a simple serial terminal emulator. Most of the time picocom. For file transfer I use mpremote. But that is not your actual issue.

**Reply by AlessandroCortese** [reply]
is it possible using Segger J-Link EDU in order to flash del .elf via Debug Port?

**Reply by robert-hh** [reply]
> is it possible using Segger J-Link EDU in order to flash del .elf via Debug Port?

That would require access to the SWD pins of the SAMD21. But as far as I can tell, these are not accessible but connected to the DEBUG CPU. But the debug CPU on the board should work like a Segger EDU device. So you could try to use it with Debugger like Ozone. when the PC is connected to the debug port.
Did you try to run the device with the target USB port and push Reset twice? That should start the default bootstrap loader.

**Reply by AlessandroCortese** [reply]
Yes i have just tried but i cannot see the drive (only with debug port i can see it)

**Reply by robert-hh** [reply]
But you said, that with the debug port the drive comes up immediately. I looked into the board's manual but did not see and hints, that one can e.g. load a .hex file to be installed in the target's flash. You should however be able to attach a debugger.

With the target port, the drive comes up if you connect the board such that it is powered and push reset twice. How fast may need some trials.

**Reply by robert-hh** [reply]
There is a PR #15942 with updated deploy instructions, telling how to install an UF2 bootloader. Then, the micropython firmware can be installed.
